File: cpp/tests/python/test_integration.py
#!/usr/bin/env python3
"""Integration tests for C++ bindings - testing all components together"""

import sys
import os
import time
import numpy as np
import math

# Add build directory to Python path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../build-release/src/bindings'))

try:
    import _claude_draw_cpp as cpp
except ImportError:
    print("Failed to import C++ bindings. Make sure to build first.")
    sys.exit(1)

def test_transform_chain():
    """Test complex transformation chains"""
    print("\n=== Transform Chain Integration ===")
    
    # Create point
    p = cpp.Point2D(10.0, 20.0)
    
    # Create transformation chain
    t1 = cpp.Transform2D.translate(5.0, 10.0)
    t2 = cpp.Transform2D.rotate(math.pi / 4)  # 45 degrees
    t3 = cpp.Transform2D.scale(2.0, 2.0)
    t4 = cpp.Transform2D.translate(-5.0, -10.0)
    
    # Compose transforms
    combined = t4 @ t3 @ t2 @ t1  # Matrix multiplication
    
    # Apply to point
    result = combined.transform_point(p)
    
    # Apply step by step for verification
    step1 = t1.transform_point(p)
    step2 = t2.transform_point(step1)
    step3 = t3.transform_point(step2)
    step4 = t4.transform_point(step3)
    
    # Should get same result
    assert abs(result.x - step4.x) < 1e-5, f"X mismatch: {result.x} vs {step4.x}"
    assert abs(result.y - step4.y) < 1e-5, f"Y mismatch: {result.y} vs {step4.y}"
    print("✓ Transform chain produces consistent results")

def test_color_compositing():
    """Test color blending with multiple layers"""
    print("\n=== Color Compositing Integration ===")
    
    # Create layer stack
    background = cpp.Color(255, 255, 255, 255)  # White
    layer1 = cpp.Color(255, 0, 0, 128)          # Semi-transparent red
    layer2 = cpp.Color(0, 255, 0, 128)          # Semi-transparent green
    layer3 = cpp.Color(0, 0, 255, 64)           # Quarter-transparent blue
    
    # Composite layers
    result1 = layer1.blend_over(background)
    result2 = layer2.blend_over(result1)
    result3 = layer3.blend_over(result2)
    
    # Final color should have components from all layers
    assert result3.r > 0, "Should have red component"
    assert result3.g > 0, "Should have green component"
    assert result3.b > 0, "Should have blue component"
    assert result3.a == 255, "Should be opaque"
    
    print(f"✓ Composited color: RGB({result3.r}, {result3.g}, {result3.b})")

def test_batch_transform_bbox():
    """Test batch operations with transforms and bounding boxes"""
    print("\n=== Batch Transform + BBox Integration ===")
    
    # Create batch of random points
    count = 10000
    points = np.random.uniform(-50, 50, size=(count, 2)).astype(np.float32)
    
    # Calculate original bounding box
    min_x, min_y = points.min(axis=0)
    max_x, max_y = points.max(axis=0)
    original_bbox = cpp.BoundingBox(min_x, min_y, max_x, max_y)
    
    print(f"Original bounds: ({min_x:.2f}, {min_y:.2f}) to ({max_x:.2f}, {max_y:.2f})")
    
    # Transform all points
    transform = cpp.Transform2D.rotate(math.pi / 3)
    transform = cpp.Transform2D.scale(1.5, 0.8) @ transform
    transform = cpp.Transform2D.translate(10, -20) @ transform
    
    cpp.batch.transform_points_batch(transform, points)
    
    # Calculate new bounding box
    min_x, min_y = points.min(axis=0)
    max_x, max_y = points.max(axis=0)
    transformed_bbox = cpp.BoundingBox(min_x, min_y, max_x, max_y)
    
    print(f"Transformed bounds: ({min_x:.2f}, {min_y:.2f}) to ({max_x:.2f}, {max_y:.2f})")
    
    # Test containment
    contains = cpp.batch.contains_points_batch(transformed_bbox, points)
    assert np.all(contains), "All points should be contained in bbox"
    print(f"✓ All {count} transformed points contained in new bbox")

def test_numpy_interop():
    """Test numpy array interoperability"""
    print("\n=== NumPy Interop Integration ===")
    
    # Create points in numpy
    count = 1000
    np_points = np.random.randn(count, 2).astype(np.float32) * 10
    
    # Convert to C++ and back
    cpp_points = []
    for i in range(count):
        cpp_points.append(cpp.Point2D(np_points[i]))
    
    # Modify through C++
    transform = cpp.Transform2D.scale(2.0, 3.0)
    for i, p in enumerate(cpp_points):
        cpp_points[i] = transform.transform_point(p)
    
    # Convert back to numpy
    result = np.zeros((count, 2), dtype=np.float32)
    for i, p in enumerate(cpp_points):
        result[i] = np.array(p, copy=False)
    
    # Verify transformation
    expected = np_points * np.array([2.0, 3.0])
    assert np.allclose(result, expected, rtol=1e-5), "Transformation mismatch"
    print("✓ NumPy arrays correctly transformed through C++")

def test_memory_efficiency():
    """Test memory-efficient batch operations"""
    print("\n=== Memory Efficiency Integration ===")
    
    # Use contiguous batches for efficiency
    point_batch = cpp.batch.PointBatch(10000)
    color_batch = cpp.batch.ColorBatch(10000)
    
    # Fill batches
    for i in range(10000):
        point_batch.add(float(i % 100), float(i % 200))
        color_batch.add_rgb(i % 256, (i * 2) % 256, (i * 3) % 256)
    
    # Get as numpy arrays (zero-copy views)
    points_array = point_batch.as_array()
    colors_array = color_batch.as_array()
    
    assert points_array.shape == (10000, 2), "Wrong points shape"
    assert colors_array.shape == (10000,), "Wrong colors shape"
    
    # Verify data integrity
    assert points_array[0, 0] == 0.0, "First point X wrong"
    assert points_array[9999, 0] == 99.0, "Last point X wrong"
    
    # Modify through numpy (should affect C++ data)
    points_array[:, 0] *= 2.0
    
    # Verify modification propagated
    modified_array = point_batch.as_array()
    assert modified_array[0, 0] == 0.0, "First point should still be 0"
    assert modified_array[1, 0] == 2.0, "Second point should be doubled"
    
    print("✓ Contiguous batches provide efficient zero-copy access")

def test_real_world_rendering():
    """Simulate a real-world rendering scenario"""
    print("\n=== Real-World Rendering Simulation ===")
    
    # Create a scene with multiple shapes
    shapes = []
    
    # Generate 50 random rectangles
    np.random.seed(42)
    for i in range(50):
        # Rectangle corners
        w = np.random.uniform(5, 20)
        h = np.random.uniform(5, 20)
        vertices = [
            cpp.Point2D(0, 0),
            cpp.Point2D(w, 0),
            cpp.Point2D(w, h),
            cpp.Point2D(0, h)
        ]
        
        # Random transform
        x = np.random.uniform(-100, 100)
        y = np.random.uniform(-100, 100)
        angle = np.random.uniform(0, 2 * math.pi)
        scale = np.random.uniform(0.5, 2.0)
        
        transform = cpp.Transform2D.translate(x, y)
        transform = cpp.Transform2D.rotate(angle) @ transform
        transform = cpp.Transform2D.scale(scale, scale) @ transform
        
        # Random color
        color = cpp.Color(
            int(np.random.randint(0, 256)),
            int(np.random.randint(0, 256)),
            int(np.random.randint(0, 256)),
            200  # Slightly transparent
        )
        
        # Calculate bounding box
        bbox = cpp.BoundingBox()
        transformed_vertices = []
        for v in vertices:
            tv = transform.transform_point(v)
            transformed_vertices.append(tv)
            bbox.expand(tv)
        
        shapes.append({
            'vertices': transformed_vertices,
            'color': color,
            'bbox': bbox
        })
    
    # Find overlapping shapes
    overlaps = 0
    for i in range(len(shapes)):
        for j in range(i + 1, len(shapes)):
            if shapes[i]['bbox'].intersects(shapes[j]['bbox']):
                overlaps += 1
    
    print(f"Found {overlaps} overlapping shape pairs")
    
    # Simulate compositing
    canvas = cpp.Color(255, 255, 255, 255)  # White background
    
    for shape in shapes:
        # In real renderer, this would be per-pixel
        canvas = shape['color'].blend_over(canvas)
    
    # Canvas should have mixed colors
    assert canvas.r < 255, "Canvas should not be pure white"
    assert canvas.g < 255, "Canvas should not be pure white"
    assert canvas.b < 255, "Canvas should not be pure white"
    
    print(f"✓ Final canvas color: RGB({canvas.r}, {canvas.g}, {canvas.b})")

def test_performance_comparison():
    """Compare C++ performance with pure Python"""
    print("\n=== Performance Comparison ===")
    
    count = 100000
    
    # Test 1: Point transformation
    points = np.random.randn(count, 2).astype(np.float32)
    transform = cpp.Transform2D.rotate(0.5)
    transform = cpp.Transform2D.scale(1.5, 2.0) @ transform
    
    # C++ batch transform
    cpp_points = points.copy()
    start = time.time()
    cpp.batch.transform_points_batch(transform, cpp_points)
    cpp_time = time.time() - start
    
    # Python equivalent (simplified)
    py_points = points.copy()
    angle = 0.5
    cos_a = math.cos(angle)
    sin_a = math.sin(angle)
    start = time.time()
    for i in range(len(py_points)):
        x, y = py_points[i]
        # Rotate
        rx = x * cos_a - y * sin_a
        ry = x * sin_a + y * cos_a
        # Scale
        py_points[i, 0] = rx * 1.5
        py_points[i, 1] = ry * 2.0
    py_time = time.time() - start
    
    print(f"Transform {count} points:")
    print(f"  C++ batch: {cpp_time*1000:.2f} ms")
    print(f"  Python loop: {py_time*1000:.2f} ms")
    print(f"  Speedup: {py_time/cpp_time:.1f}x")
    
    # Test 2: Distance calculation
    points1 = np.random.randn(count//10, 2).astype(np.float32)
    points2 = np.random.randn(count//10, 2).astype(np.float32)
    
    # C++ batch distance
    start = time.time()
    cpp_distances = cpp.batch.calculate_distances_batch(points1, points2)
    cpp_time = time.time() - start
    
    # NumPy distance
    start = time.time()
    np_distances = np.sqrt(((points1 - points2) ** 2).sum(axis=1))
    np_time = time.time() - start
    
    print(f"\nCalculate {count//10} distances:")
    print(f"  C++ batch: {cpp_time*1000:.2f} ms")
    print(f"  NumPy: {np_time*1000:.2f} ms")
    
    # Verify correctness
    assert np.allclose(cpp_distances, np_distances, rtol=1e-5), "Distance mismatch"

def test_edge_cases():
    """Test edge cases and error handling"""
    print("\n=== Edge Cases Integration ===")
    
    # Empty bounding box
    empty_box = cpp.BoundingBox()
    assert empty_box.is_empty(), "Default bbox should be empty"
    
    # Degenerate transform
    degenerate = cpp.Transform2D.scale(0.0, 1.0)
    assert not degenerate.is_invertible(), "Should not be invertible"
    
    # Zero-alpha blending
    transparent = cpp.Color(255, 0, 0, 0)
    opaque = cpp.Color(0, 0, 255, 255)
    result = transparent.blend_over(opaque)
    assert result.rgba == opaque.rgba, "Transparent should have no effect"
    
    # Very small distances
    p1 = cpp.Point2D(0.0, 0.0)
    p2 = cpp.Point2D(1e-7, 1e-7)
    dist = p1.distance_to(p2)
    assert dist > 0, "Distance should be positive"
    assert dist < 1e-6, "Distance should be very small"
    
    # Large batch allocation
    try:
        huge_batch = cpp.batch.create_points(1000000, 0, 0)
        assert huge_batch.shape[0] == 1000000, "Should create 1M points"
        print("✓ Successfully allocated 1M points")
    except MemoryError:
        print("✓ Large allocation failed gracefully")
    
    print("✓ All edge cases handled correctly")

def main():
    """Run all integration tests"""
    print("=== C++ Integration Tests ===")
    
    test_transform_chain()
    test_color_compositing()
    test_batch_transform_bbox()
    test_numpy_interop()
    test_memory_efficiency()
    test_real_world_rendering()
    test_performance_comparison()
    test_edge_cases()
    
    print("\n✅ All integration tests passed!")

if __name__ == "__main__":
    main()

File: cpp/tests/integration/test_integration.cpp
#include <gtest/gtest.h>
#include <vector>
#include <random>
#include <chrono>

#include "claude_draw/core/point2d.h"
#include "claude_draw/core/color.h"
#include "claude_draw/core/transform2d.h"
#include "claude_draw/core/bounding_box.h"
#include "claude_draw/memory/type_pools.h"
#include "claude_draw/batch/batch_operations.h"

using namespace claude_draw;
using namespace claude_draw::memory;
using namespace claude_draw::batch;

class IntegrationTest : public ::testing::Test {
protected:
    void SetUp() override {
        // Clear pools before each test
        TypePools::instance().clear_all();
    }
};

TEST_F(IntegrationTest, TransformChain) {
    // Test complex transformation chain
    Point2D p(10.0f, 20.0f);
    
    // Create transformation chain
    Transform2D t1 = Transform2D::translate(5.0f, 10.0f);
    Transform2D t2 = Transform2D::rotate(M_PI / 4);  // 45 degrees
    Transform2D t3 = Transform2D::scale(2.0f, 2.0f);
    Transform2D t4 = Transform2D::translate(-5.0f, -10.0f);
    
    // Compose transforms
    Transform2D combined = t4 * t3 * t2 * t1;
    
    // Apply to point
    Point2D result = combined.transform_point(p);
    
    // Apply step by step for verification
    Point2D step1 = t1.transform_point(p);
    Point2D step2 = t2.transform_point(step1);
    Point2D step3 = t3.transform_point(step2);
    Point2D step4 = t4.transform_point(step3);
    
    // Should get same result
    EXPECT_NEAR(result.x, step4.x, 1e-5f);
    EXPECT_NEAR(result.y, step4.y, 1e-5f);
}

TEST_F(IntegrationTest, ColorBlendingChain) {
    // Test multiple color blending operations
    Color background(255, 255, 255, 255);  // White
    Color layer1(255, 0, 0, 128);          // Semi-transparent red
    Color layer2(0, 255, 0, 128);          // Semi-transparent green
    Color layer3(0, 0, 255, 64);           // Quarter-transparent blue
    
    // Blend layers
    Color result1 = layer1.blend_over(background);
    Color result2 = layer2.blend_over(result1);
    Color result3 = layer3.blend_over(result2);
    
    // Final color should have components from all layers
    EXPECT_GT(result3.r, 0);
    EXPECT_GT(result3.g, 0);
    EXPECT_GT(result3.b, 0);
    EXPECT_EQ(result3.a, 255);  // Should be opaque
}

TEST_F(IntegrationTest, BoundingBoxTransformations) {
    // Test bounding box with transformations
    BoundingBox box(0.0f, 0.0f, 10.0f, 10.0f);
    
    // Get corners
    Point2D corners[4] = {
        Point2D(box.min_x, box.min_y),
        Point2D(box.max_x, box.min_y),
        Point2D(box.max_x, box.max_y),
        Point2D(box.min_x, box.max_y)
    };
    
    // Transform corners
    Transform2D transform = Transform2D::rotate(M_PI / 6) * Transform2D::scale(2.0f, 1.5f);
    
    BoundingBox transformed_box;
    for (int i = 0; i < 4; ++i) {
        Point2D transformed = transform.transform_point(corners[i]);
        transformed_box.expand(transformed);
    }
    
    // Verify transformed box contains all transformed corners
    for (int i = 0; i < 4; ++i) {
        Point2D transformed = transform.transform_point(corners[i]);
        EXPECT_TRUE(transformed_box.contains(transformed));
    }
}

TEST_F(IntegrationTest, MemoryPoolIntegration) {
    // Test using memory pools with batch operations
    const size_t count = 10000;
    
    // Pre-allocate pools
    TypePools& pools = TypePools::instance();
    pools.reserve_points(count);
    pools.reserve_colors(count);
    
    // Create objects using pools
    std::vector<PooledPtr<Point2D, TypePools::POINT_BLOCK_SIZE>> points;
    std::vector<PooledPtr<Color, TypePools::COLOR_BLOCK_SIZE>> colors;
    
    points.reserve(count);
    colors.reserve(count);
    
    for (size_t i = 0; i < count; ++i) {
        points.push_back(TypePools::make_point(
            static_cast<float>(i % 100), 
            static_cast<float>(i % 200)
        ));
        colors.push_back(TypePools::make_color(
            static_cast<uint8_t>(i % 256),
            static_cast<uint8_t>((i * 2) % 256),
            static_cast<uint8_t>((i * 3) % 256),
            255
        ));
    }
    
    // Verify pool usage
    auto point_stats = pools.point_stats();
    auto color_stats = pools.color_stats();
    
    EXPECT_GE(point_stats.allocated, count);
    EXPECT_GE(color_stats.allocated, count);
}

TEST_F(IntegrationTest, BatchTransformIntegration) {
    // Test batch operations with transforms and bounding boxes
    const size_t count = 1000;
    
    // Create batch of points
    ContiguousBatch<Point2D> points(count);
    std::uniform_real_distribution<float> dist(-50.0f, 50.0f);
    std::mt19937 rng(42);
    
    for (size_t i = 0; i < count; ++i) {
        points.emplace(dist(rng), dist(rng));
    }
    
    // Create bounding box from points
    BoundingBox original_bbox;
    for (const auto& p : points.as_span()) {
        original_bbox.expand(p);
    }
    
    // Transform all points
    Transform2D transform = Transform2D::rotate(M_PI / 3) * 
                           Transform2D::scale(1.5f, 0.8f) *
                           Transform2D::translate(10.0f, -20.0f);
    
    BatchProcessor::transform_points(transform, points.as_span());
    
    // Create new bounding box
    BoundingBox transformed_bbox;
    for (const auto& p : points.as_span()) {
        transformed_bbox.expand(p);
    }
    
    // Test containment
    std::vector<bool> contains(count);
    BatchProcessor::contains_points(transformed_bbox, points.data(), contains.data(), count);
    
    // All points should be contained
    for (size_t i = 0; i < count; ++i) {
        EXPECT_TRUE(contains[i]) << "Point " << i << " not contained";
    }
}

TEST_F(IntegrationTest, RealWorldScenario) {
    // Simulate a real-world graphics operation
    // Drawing multiple colored shapes with transformations
    
    struct Shape {
        std::vector<Point2D> vertices;
        Color color;
        Transform2D transform;
        BoundingBox bounds;
    };
    
    // Create shapes
    std::vector<Shape> shapes;
    shapes.reserve(100);
    
    // Create 100 random rectangles
    std::mt19937 rng(42);
    std::uniform_real_distribution<float> pos_dist(-100.0f, 100.0f);
    std::uniform_real_distribution<float> size_dist(5.0f, 20.0f);
    std::uniform_real_distribution<float> angle_dist(0.0f, 2 * M_PI);
    std::uniform_int_distribution<uint8_t> color_dist(0, 255);
    
    for (int i = 0; i < 100; ++i) {
        Shape shape;
        
        // Rectangle vertices
        float w = size_dist(rng);
        float h = size_dist(rng);
        shape.vertices = {
            Point2D(0, 0),
            Point2D(w, 0),
            Point2D(w, h),
            Point2D(0, h)
        };
        
        // Random color
        shape.color = Color(
            color_dist(rng),
            color_dist(rng),
            color_dist(rng),
            200  // Slightly transparent
        );
        
        // Random transform
        shape.transform = Transform2D::translate(pos_dist(rng), pos_dist(rng)) *
                         Transform2D::rotate(angle_dist(rng)) *
                         Transform2D::scale(
                             std::uniform_real_distribution<float>(0.5f, 2.0f)(rng),
                             std::uniform_real_distribution<float>(0.5f, 2.0f)(rng)
                         );
        
        // Calculate bounds
        shape.bounds = BoundingBox();
        for (const auto& v : shape.vertices) {
            Point2D transformed = shape.transform.transform_point(v);
            shape.bounds.expand(transformed);
        }
        
        shapes.push_back(shape);
    }
    
    // Find overlapping shapes
    int overlaps = 0;
    for (size_t i = 0; i < shapes.size(); ++i) {
        for (size_t j = i + 1; j < shapes.size(); ++j) {
            if (shapes[i].bounds.intersects(shapes[j].bounds)) {
                overlaps++;
            }
        }
    }
    
    std::cout << "Found " << overlaps << " overlapping shape pairs" << std::endl;
    
    // Composite all shapes
    Color canvas(255, 255, 255, 255);  // White background
    
    // Simple compositing (would be per-pixel in real renderer)
    for (const auto& shape : shapes) {
        canvas = shape.color.blend_over(canvas);
    }
    
    // Canvas should have mixed colors
    EXPECT_LT(canvas.r, 255);
    EXPECT_LT(canvas.g, 255);
    EXPECT_LT(canvas.b, 255);
}

TEST_F(IntegrationTest, PerformanceStressTest) {
    // Stress test with large number of operations
    const size_t object_count = 100000;
    const size_t iteration_count = 10;
    
    auto start = std::chrono::high_resolution_clock::now();
    
    for (size_t iter = 0; iter < iteration_count; ++iter) {
        // Create objects
        auto points = BatchCreator::create_points(object_count, 0.0f, 0.0f);
        auto colors = BatchCreator::create_colors(object_count, 0xFF000000);
        
        // Transform points
        Transform2D transform = Transform2D::rotate(0.1f) * Transform2D::scale(1.1f, 0.9f);
        
        for (auto* p : points) {
            *p = transform.transform_point(*p);
        }
        
        // Blend colors
        Color background(255, 255, 255, 255);
        for (auto* c : colors) {
            *c = c->blend_over(background);
        }
        
        // Calculate bounding box
        BoundingBox bbox;
        for (auto* p : points) {
            bbox.expand(*p);
        }
        
        // Clean up
        BatchCreator::destroy_batch(points);
        BatchCreator::destroy_batch(colors);
    }
    
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    
    std::cout << "Stress test completed in " << duration.count() << " ms" << std::endl;
    std::cout << "Average per iteration: " << duration.count() / iteration_count << " ms" << std::endl;
    
    // Should complete in reasonable time
    EXPECT_LT(duration.count(), 10000);  // Less than 10 seconds
}

TEST_F(IntegrationTest, EdgeCasesAndErrorHandling) {
    // Test edge cases with all components
    
    // Empty bounding box
    BoundingBox empty_box;
    EXPECT_TRUE(empty_box.is_empty());
    
    // Transform empty box corners
    Transform2D t = Transform2D::scale(2.0f, 2.0f);
    Point2D p(empty_box.min_x, empty_box.min_y);
    Point2D transformed = t.transform_point(p);
    // Should handle gracefully
    
    // Degenerate transform
    Transform2D degenerate = Transform2D::scale(0.0f, 1.0f);
    EXPECT_FALSE(degenerate.is_invertible());
    
    // Zero-alpha color blending
    Color transparent(255, 0, 0, 0);
    Color opaque(0, 0, 255, 255);
    Color result = transparent.blend_over(opaque);
    EXPECT_EQ(result.rgba, opaque.rgba);  // Transparent has no effect
    
    // Very small distances
    Point2D p1(0.0f, 0.0f);
    Point2D p2(1e-7f, 1e-7f);
    float dist = p1.distance_to(p2);
    EXPECT_GT(dist, 0.0f);
    EXPECT_LT(dist, 1e-6f);
    
    // Large batch operations
    const size_t huge_count = 1000000;
    try {
        auto huge_batch = BatchCreator::create_points(huge_count, 0.0f, 0.0f);
        EXPECT_EQ(huge_batch.size(), huge_count);
        BatchCreator::destroy_batch(huge_batch);
    } catch (const std::bad_alloc&) {
        // Acceptable if system runs out of memory
        std::cout << "Large allocation failed (expected on low-memory systems)" << std::endl;
    }
}